Stirng not null terminated

nishith-vihar opened the pull request:
CID: 1214629,1274235,1430115,1437648

Null character is added at the end of buffer which corrects the
issue.

Change-Id: I8f7016520ffd41b2c68fe3c7f053e0e04f306c84
Updates: #1060
Signed-off-by: Nishith Vihar Sakinala <nsakinal@redhat.com>

nishith-vihar committed:
Stirng not null terminated
CID: 1214629,1274235,1430115,1437648

Null character is added at the end of buffer which corrects the
issue.

Change-Id: I8f7016520ffd41b2c68fe3c7f053e0e04f306c84
Updates: #1060
Signed-off-by: Nishith Vihar Sakinala <nsakinal@redhat.com>

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

mykaul commented on libglusterfs/src/syscall.c:
```diff
@@ -362,7 +364,9 @@ sys_writev(int fd, const struct iovec *iov, int iovcnt)
 ssize_t
 sys_read(int fd, void *buf, size_t count)
```
Who says we only read strings here? What if it was binary data?

nishith-vihar replied:
Sorry. Actually the coverity issue mentions it as "String not null terminated". I guess the null character should be added in the function calling sys_read which passes a char buffer as argument. And i guess in the case of sys_readlink null character can be appended right?

mykaul replied:
Correct. 'man 2 readlink' indeed says:
```
readlink() does not append  a  null byte  to buf.  It will (silently) truncate the contents (to a length of
bufsiz characters), in case the buffer is too small to hold all of  the contents.
```

nishith-vihar committed:
String not null terminated
CID: 1430115

The buffer in sys_readlink is not terminated by null.
So the null is appended in the buffer

Change-Id: I5501c8adc093bd920abdf09590d518408f3286da
Updates: #1060
Signed-off-by: Nishith Vihar Sakinala <nsakinal@redhat.com>

xhernandez commented on libglusterfs/src/syscall.c:
```diff
@@ -187,7 +187,9 @@ sys_readdir(DIR *dir, struct dirent *de)
 {
-    return FS_RET_CHECK(readlink(path, buf, bufsiz), errno);
+    int sz = readlink(path, buf, bufsiz);
+    buf[sz] = '\0';
```
In case of error `sz` will be -1, so this is incorrect.

xhernandez commented on libglusterfs/src/syscall.c:
```diff
@@ -187,7 +187,9 @@ sys_readdir(DIR *dir, struct dirent *de)
 sys_readlink(const char *path, char *buf, size_t bufsiz)
 {
-    return FS_RET_CHECK(readlink(path, buf, bufsiz), errno);
+    int sz = readlink(path, buf, bufsiz);
```
This is a simple wrapper of the `readlink()` system call. The behavior of this function should be exactly the same as calling `readlink()` directly. What should need to be updated are the callers of this function.

nishith-vihar replied:
Thanks for the suggestion. I have made the required changes and have added the null character at the caller of the function.

nishith-vihar committed:
String not null terminated
CID: 1430115

The buffer is not terminated by null after calling sys_readlink.
So the null is appended in the buffer.

Change-Id: I47099f2b5943e53418083d4f2ecf4145ed9736f5
Updates: #1060
Signed-off-by: Nishith Vihar Sakinala <nsakinal@redhat.com>

nik-redhat commented:
/run regression

xhernandez commented on geo-replication/src/gsyncd.c:
```diff
@@ -10,11 +10,11 @@
-#include <stdlib.h>
 #include <stdio.h>
-#include <unistd.h>
+#include <stdlib.h>
```
Any reason for changing the order of the includes ?

nishith-vihar replied:
I suppose when we run clang-format on the file it rearranges it accordingly. That's all

xhernandez replied:
Current `.clang-format` configuration has `SortIncludes:    false`, so it shouldn't have happened. Can you try resubmit the patch without the changes in the include files ? it's better to keep changes as small as possible.

nishith-vihar replied:
I have made those changes and pushed it again and this time the clang format test is failing because of the rearranged includes

xhernandez replied:
I don't understand why it happens. Anyway, sorry for the noise. If you can resubmit your original patch I'll accept it.

schaffung replied:
Is it because of change in clang version ? ( was mentioned in the community meeting ) I think a new clang version was integrated with the CI and maybe that is causing it to fail.

nishith-vihar replied:
> I don't understand why it happens. Anyway, sorry for the noise. If you can resubmit your original patch I'll accept it.

Thank you. I have resubmitted the patch.

schaffung commented:
/run regression
